bundles/org.openhab.binding.enocean/src/main/java/org/openhab/binding/enocean/internal/eep/A5_09/A5_09_04.java

```diff
@@ -0,0 +1,121 @@
+/**
+ * Copyright (c) 2010-2022 Contributors to the openHAB project
+ *
+ * See the NOTICE file(s) distributed with this work for additional
+ * information.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ */
+package org.openhab.binding.enocean.internal.eep.A5_09;
+
+import static org.openhab.binding.enocean.internal.EnOceanBindingConstants.*;
+
+import java.util.function.Function;
+
+import org.openhab.binding.enocean.internal.messages.ERP1Message;
+import org.openhab.core.config.core.Configuration;
+import org.openhab.core.library.types.DecimalType;
+import org.openhab.core.library.types.QuantityType;
+import org.openhab.core.library.unit.SIUnits;
+import org.openhab.core.library.unit.Units;
+import org.openhab.core.types.State;
+import org.openhab.core.types.UnDefType;
+
+/**
+ *
+ * @author Zhivka Dimova - Initial contribution
+ */
+public class A5_09_04 extends A5_09 {
+
+    public A5_09_04() {
+        super();
+    }
+
+    public A5_09_04(ERP1Message packet) {
+        super(packet);
+    }
+
+    protected double getUnscaledCO2Min() {
+        return 0;
+    }
+
+    protected double getUnscaledCO2Max() {
+        return 255;
+    }
+
+    protected double getScaledCO2Min() {
+        return 0;
+    }
+
+    protected double getScaledCO2Max() {
+        return 2550;
+    }
+
+    protected int getUnscaledCO2Value() {
+        return getDB_2Value();
+    }
+
+    protected double getUnscaledTemperatureMin() {
+        return 0;
+    }
+
+    protected double getUnscaledTemperatureMax() {
+        return 255;
+    }
+
+    protected double getScaledTemperatureMin() {
+        return 0;
+    }
+
+    protected double getScaledTemperatureMax() {
+        return 51;
+    }
+
+    protected int getUnscaledTemperatureValue() {
+        return getDB_1Value();
+    }
+
+    protected double getUnscaledHumidityMin() {
+        return 0;
+    }
+
+    protected double getUnscaledHumidityMax() {
+        return 200;
+    }
+
+    protected int getUnscaledHumidityValue() {
+        return getDB_3Value();
+    }
+
+    protected double getScaledHumidityMin() {
+        return 0;
+    }
+
+    protected double getScaledHumidityMax() {
+        return 100;
+    }
+
+    @Override
+    protected State convertToStateImpl(String channelId, String channelTypeId,
+            Function<String, State> getCurrentStateFunc, Configuration config) {
+
+        if (channelId.equals(CHANNEL_CO2)) {
+            double scaledCO2 = getScaledCO2Min() + ((getUnscaledCO2Value() * (getScaledCO2Max() - getScaledCO2Min()))
```

Comment:
This cumbersome scaling mechanism is just needed if you implement the whole famliy A5-09 family. In this case you add this scaling function to the base class and override the scaling factors in the derived class. As you have just implemented the EEP A5-09-04 I would be fine if you just multipy Byte2 with 10. This would improve the readability. If you want to keep this conversion, I would suggest to move this conversion to the base class.

Reply:
@lovery - just checking, did you see this comment?

Reply:
@jlaur did you ment to mention @fruggy83 or me? The last comment is by me, asking @fruggy83 if the above code is what he meant in his comment.

Reply:
I think I meant to tag @fruggy83 here.

Reply:
@lovery if it is the same, I would put it in the base class. You could look at the implementation of the A5_08 EEP family. All subclasses just override the scale factor functions. None of them implements the convertToStateImpl function.
A5_12 is another example. Here the subclasses even return different quantity types.

Reply:
@lovery @jlaur 
If you plan to release a new version in the next few days, I would be also fine with the current implementation and creating another issue for the refactoring part. This is up to you @jlaur 😃

Reply:
@fruggy83 - unfortunately it's too late for 3.4, so if you and @lovery agree to refactor this, there will be no rush. So actually it would be up to you, @lovery, if you prefer to have this merged soon and do a follow-up PR, or if you'd rather finish within the context of this one?

Reply:
@fruggy83 @jlaur I prefer to finish in this PR and not open a new one for refactor. I also started an implementation for A5_09_02, A5_09_08, A5_09_09, A5_09_0C if you don't mind I will include it in this PR